### ISSUE TITLE
Flag attribute enhancements

### ISSUE THREAD

**Issue by xatapult** [Schematron/schematron-enhancement-proposals#12]
A `flag` attribute is now typed as a simple `xs:string`. But it is supposed to be the name of a *variable*...

So wouldn't it be better to type it as `xs:NCName`? Or some other restrictive type?

Even better: I see no reason why a triggered assert, report or rule couldn't raise multiple flags... So I would propose to type the `flag` attribute as a list of its base type.

**Comment by rjelliffe** [comment]
[COMMENT WITHDRAWN]

As to an assert etc raising multiple flags, that would not change the idea
of a flag.

As to the value of a flag being the ids of the asserts etc that raise it,
rather than it just being raised or not raised, that would not change the
idea of a flag (just provide independent tracing).  (The reason a flag is
not boolean is that absence means not raised.)

But an implementation can already provide this information currently, with
a foreign element or attribute on the SVRL... if none has, can we say there
is demand or experience?

As to which xsd data type it should be, my personal preference is to have
as little to do with XSD as possible. Indeed, I would prefer if only DTDs
were used. But it is not my call.

Flags allow  feature detection in a document, especially to guide
subsequent processing or classification.

Regards
Rick~~

**Comment by xatapult** [comment]
@rjelliffe 

>As to an assert etc raising multiple flags, that would not change the idea
of a flag.

Nope. But we could allow this. The specification now specifically mentions "boolean variable" (singular).

>As to the value of a flag being the ids of the asserts etc that raise it,
rather than it just being raised or not raised, that would not change the
idea of a flag (just provide independent tracing).  (The reason a flag is
not boolean is that absence means not raised.)

Flags are boolean variables according to the spec, so I'm not sure what you mean

I'm curious what the reasons are to be against strong datatyping (whatever language used, Schema. RelaxNG, whatever).

**Comment by rjelliffe** [comment]
[COMMENT WITHDRAWN]

I have a field report!

A client has a big Schematron implementation as part of their multi-user CMS. They used @flag where they should have used @role, to pass around severity levels.  They did not notice @role and it was convenient in the SVRL.

ORIGINAL INTENT OF FLAG
---------------------------------------

Now, to be strict for the original intent of @flag, it was that a perfect version of SVRL would have something like

<svrl:schematron-report>
   <svrl:flag-raised flag="invalid"/>
   <svrl:flag-raised flag="has-severe-error" />
   <svrl:flag-raised flag="uses-version-27-features"/>
   <svrl:flag-raised flag="not-dispatchable" />
...

E.g. a flag is a property of the whole document, and of the whole validation session. It is a summary or simple property: if some assertion has failed, or some report has succeeded, what does it mean for the whole document? 

The reason that individual failed-asserts have their own flag is 

1) to allow convenient location, 
2) extra internal passes of the SVRL wasnot really possible for XSLT 1, and 
3) I didn't want to put in the extra external pass to summarize the SVRL flags (though IIRC I put in notes mentioning that to extra the flag status you need to look for any SVRL item having that flag on it. 

Now this is why a flag was originally defined as not being a boolean but as something that existed or not.  

But I guess that since the standard now treats it as boolean, you would be having something like:

<svrl:schematron-report>
  <scrl:flags-summary>
      <svrl:flag name="invalid"  value="true" />
      <svrl:flag name="has-severe-error"  value="true" />
      <svrl:flag name="uses-version-27-features"  value="true" />
      <svrl:flag name="not-dispatchable"   value="true" />
      <svrl:flag name="urgent-issue" value="false" />
      ...
   </svrl:flag-summary>


SUGGESTION

I do not see any need for a flag identifier to be a simple token type rather than a string: let the user choose.

But I think there is a gap in the specification that confuses that while we do want our svrl:failed-assert to signal that a flag was raised by it, there is no top-level place to show flags.  Putting this in a separate page.

I think the best approach is to add to SVRL something like the <svrl:flag-raised  element as above, with wording that the usual XPath rule of absence=false() applies.

**Comment by rjelliffe** [comment]
Can we approach this another way?

1) Yes, define that `sch:*/@flag` and `svrl:*/@flag`  are lists of tokens e.g. Name. This was the OP's request. 

2) Also, define that there are two separate things, the "report flags" which is the value of the flags attribute on an svrl:*  element, and the "document flags" which are the concatenation of all svrl:*/@flags values (unique-ized) in the SVRL report.  This was what I was concerned about.

3) Also, preferably, clarify that flags are not boolean types (i.e. true or false) nor xs:boolean types (i.e. "true", "false", "0", "1"), but named properties which can be spoken of as "raised"  if detected or "not raised" otherwise.  (For developers who want to understand them in programming terms, they are more like enumeration values: the name is the value.)

**Comment by AndrewSales** [comment]
> But I guess that since the standard now treats it as boolean

Just noting here that my copy of the FDIS from 2004 defines a flag as a boolean. So I think that in ISO Schematron, it always was so.

**Comment by AndrewSales** [comment]
> named properties which can be spoken of as "raised" if detected or "not raised" otherwise

Yes. My interpretation is that a flag just means "this thing I am interested in occurred". I think the wording in the standard is a bit unfortunate, because "variable" implies something else, particularly as it's defined formally in the normative terms and definitions -- perhaps even "boolean value" would have been clearer.

I agree it makes sense for this attribute's datatype to become a list of tokens. Strictly speaking, this is (subtly) backward-incompatible, but does not affect the outcome of validation, only its reporting. In any case, any reliance on the difference in whitespace handling between `string` and `token` is likely very unusual and probably unwise.

**Comment by rjelliffe** [comment]
> I agree it makes sense for this attribute's datatype to become a list of
tokens.

Yes.

A flag is not a binary value (e.g., a boolean, [1|0], [true|false]) but a
unary value  (i.e., the value is a set which either has the token present
or is empty ∅.)  Similarly, the ***@***.*** result is a set which either
has one or more tokens or is empty ∅.

Regards
Rick

**Comment by AndrewSales** [comment]
I think this now meets your criteria, @xatapult, so closing.
